File: server/data_server/logic/logic.py
""" logic.py

This module contains all methods to perform any business logic required
between the presentation and data layers. This is the middle layer in the
3-tier architecture used in this project.

This layer is primarily used to do any data conversion between the expectation
at the presentation layer and the expectation at the data layer.
"""
import os
import json
import warnings
from email.mime.text import MIMEText
from email.mime.multipart import MIMEMultipart
from dateutil import tz
from datetime import datetime
from six import PY2, PY3
import data_server.shared.collection_names as collections
from data_server.data.sqlite_data import SqliteData

_data = SqliteData()


def connect(location):
    """ connect method

    Attempts to create a connection with the given DB at the given
    location for SQLite files

    Returns:
        bool: True if connection established, False otherwise
    """
    return _data.connect(location)

def disconnect():
    """ disconnect method

    Disconnect the data layer.

    Returns:
        bool: True if disconnected, False otherwise. Defaults to True
    """
    if _data is not None:
        return _data.disconnect()
    return True


def update_layout(body):
    """ update_layout method

    Updates the layout for a given case.

    Args:
        body (JSON): the new layout
    Returns:
        True if success, False otherwies
    """
    return _data.update_layout(json.dumps(body))


def metadata_get():
    """ metadata_get method

    Grabs and returns the metadata for a given case
    """
    n_abs2prom = {'input': {}, 'output': {}}
    n_prom2abs = {'input': {}, 'output': {}}

    res = _data.generic_get(collections.METADATA, False)

    if res is not None and 'abs2prom' in res:
        for io in ['input', 'output']:
            for k in res['abs2prom'][io]:
                n_k = k.replace('___', '.')
                n_abs2prom[io][n_k] = res['abs2prom'][io][k]

            for k in res['prom2abs'][io]:
                n_k = k.replace('___', '.')
                n_prom2abs[io][n_k] = res['prom2abs'][io][k]

        res['abs2prom'] = n_abs2prom
        res['prom2abs'] = n_prom2abs

        return json.dumps(res)
    else:
        print("No metadata stored")
        return "{}"


def generic_get(collection_name, get_many=True):
    """ generic_get method

    Performs the typical 'get' request, returning the list of documents in the given collection.

    Args:
        collection_name (string): the collection to be queried
        get_many (bool): true if you want to get all, false if you only want
        one
    Returns:
        Array of documents
    """
    return json.dumps(_data.generic_get(collection_name, get_many))


def get_driver_iteration_data(variable):
    """ get_driver_iteration_data method

    Grabs and returns all data for a given variable over each iteration
    in the driver_iteration collection

    Args:
        variable (string): the variable whose data is to be used for querying
    Returns:
        Array of data
    """
    dat = _data.get_driver_iteration_data()
    ret = []
    for i in dat:
        for v in i['desvars']:
            if v['name'] == variable:
                v['iteration'] = _extract_iteration_coordinate(
                    i['iteration_coordinate'])
                v['counter'] = i['counter']
                v['type'] = 'desvar'
                ret.append(v)
        for v in i['objectives']:
            if v['name'] == variable:
                v['iteration'] = _extract_iteration_coordinate(
                    i['iteration_coordinate'])
                v['counter'] = i['counter']
                v['type'] = 'objective'
                ret.append(v)
        for v in i['constraints']:
            if v['name'] == variable:
                v['iteration'] = _extract_iteration_coordinate(
                    i['iteration_coordinate'])
                v['counter'] = i['counter']
                v['type'] = 'constraint'
                ret.append(v)

        if 'sysincludes' in i:
            for v in i['sysincludes']:
                if v['name'] == variable:
                    v['iteration'] = _extract_iteration_coordinate(
                        i['iteration_coordinate'])
                    v['counter'] = i['counter']
                    v['type'] = 'sysinclude'
                    ret.append(v)

        if 'inputs' in i:
            for v in i['inputs']:
                if v['name'] == variable:
                    v['iteration'] = _extract_iteration_coordinate(
                        i['iteration_coordinate'])
                    v['counter'] = i['counter']
                    v['type'] = 'constraint'
                    ret.append(v)

    return json.dumps(ret)


def get_all_driver_vars():
    """ get_all_driver_vars method

    Grabs all variables in driver_iterations

    Returns:
        Array of string variable names
    """
    dat = _data.get_driver_iteration_data()
    ret = []
    cache = []
    for i in dat:
        for v in i['desvars']:
            if v['name'] not in cache:
                ret.append({
                    'name': v['name'],
                    'type': 'desvar'
                })
                cache.append(v['name'])
        for v in i['objectives']:
            if v['name'] not in cache:
                ret.append({
                    'name': v['name'],
                    'type': 'objective'
                })
                cache.append(v['name'])
        for v in i['constraints']:
            if v['name'] not in cache:
                ret.append({
                    'name': v['name'],
                    'type': 'constraint'
                })
                cache.append(v['name'])

        if 'sysincludes' in i:
            for v in i['sysincludes']:
                if v['name'] not in cache:
                    ret.append({
                        'name': v['name'],
                        'type': 'sysinclude'
                    })
                    cache.append(v['name'])

        if 'inputs' in i:
            for v in i['inputs']:
                if v['name'] not in cache:
                    ret.append({
                        'name': v['name'],
                        'type': 'constraint'
                    })
                    cache.append(v['name'])

    return json.dumps(ret)


def get_driver_iteration_based_on_count(variable, count):
    """ get_highest_driver_iteartion_count method

    Returns data if new data is available (checked by count)

    Args:
        variable (string): the variable to be checked
        count (int): the max count up to this point
    Returns:
        JSON string of the data or of '[]' if no update necessary
    """
    if _data.is_new_data(count):
        s_data = get_driver_iteration_data(variable)
        return s_data

    return "[]"


def _extract_iteration_coordinate(coord):
    """ private extract_iteration_coordinate method

    Splits up the iteration coordinate string and returns it as a JSON array

    Args:
        coord (string): the coordinate to be split up
    Returns:
        Array of JSON of the object
    """
    delimiter = '|'
    ret = []

    split_coord = coord.split(delimiter)
    i = 0
    while i < len(split_coord):
        node = {}
        node['name'] = split_coord[i]
        node['iteration'] = split_coord[i + 1]
        ret.append(node)
        i += 2

    return ret

File: src/data_server/logic/logic.js
'use strict';

const SqliteData = require('../data/sqlite_data');

class Logic {
    constructor(logger) {
        this._logger = logger;
        this._data = new SqliteData(logger);
    }

    /**
     * Connect to DB with the given filename
     *
     * @param {string} fname
     * @returns {Promise}
     */
    async connect(fname) {
        return this._data.connect(fname);
    }

    /**
     * Disconnect from the database
     */
    disconnect() {
        if (this._data != null) {
            this._data.disconnect();
        }
    }

    /**
     * Store this layout as the current layout
     *
     * @param {*} layout
     * @returns {Promise}
     */
    async updateLayout(layout) {
        return this._data.updateLayout(JSON.stringify(layout));
    }

    /**
     * Get the current layout
     *
     * @returns {Promise}
     */
    async getLayout() {
        return this._data.getLayout();
    }

    /**
     * Get abs2meta, abs2prom and prom2abs from database
     *
     * @returns {Promise} resolves to metadata
     */
    async getMetadata() {
        let n_abs2prom = { input: {}, output: {} };
        let n_prom2abs = { input: {}, output: {} };

        let ret = await this._data.getMetadata();
        if (ret != null && ret.hasOwnProperty('abs2prom')) {
            let io = ['input', 'output'];
            for (let i = 0; i < io.length; ++i) {
                let inp_out = io[i];
                for (
                    let j = 0;
                    j < Object.keys(ret['abs2prom'][inp_out]).length;
                    ++j
                ) {
                    let v = Object.keys(ret['abs2prom'][inp_out])[j];
                    let n_k = v.replace('___', '.');
                    n_abs2prom[inp_out][n_k] = ret['abs2prom'][inp_out][v];
                }
                for (
                    let j = 0;
                    j < Object.keys(ret['prom2abs'][inp_out]).length;
                    ++j
                ) {
                    let v = Object.keys(ret['prom2abs'][inp_out])[j];
                    let n_k = v.replace('___', '.');
                    n_prom2abs[inp_out][n_k] = ret['prom2abs'][inp_out][v];
                }
            }

            return { abs2prom: n_abs2prom, prom2abs: n_prom2abs };
        } else {
            this._logger.warn(
                'Trouble getting metadata. Error: ' + ret.toString()
            );
            return {};
        }
    }

    /**
     * Get all driver iteration data for a given variable
     *
     * @param {string} variable
     * @returns {Promise} resolves to iteration data
     */
    async getDriverIterationData(variable) {
        let self = this;
        let dat = await this._data.getDriverIterationData();
        let ret = [];
        dat.forEach(i => {
            i['desvars'].forEach(v => {
                if (v['name'] === variable) {
                    v['iteration'] = self._exractIterationCoordinate(
                        i['iteration_coordinate']
                    );
                    v['counter'] = i['counter'];
                    v['type'] = 'desvar';
                    ret.push(v);
                }
            });
            i['objectives'].forEach(v => {
                if (v['name'] === variable) {
                    v['iteration'] = self._exractIterationCoordinate(
                        i['iteration_coordinate']
                    );
                    v['counter'] = i['counter'];
                    v['type'] = 'objective';
                    ret.push(v);
                }
            });
            i['constraints'].forEach(v => {
                if (v['name'] === variable) {
                    v['iteration'] = self._exractIterationCoordinate(
                        i['iteration_coordinate']
                    );
                    v['counter'] = i['counter'];
                    v['type'] = 'constraint';
                    ret.push(v);
                }
            });
            i['sysincludes'].forEach(v => {
                if (v['name'] === variable) {
                    v['iteration'] = self._exractIterationCoordinate(
                        i['iteration_coordinate']
                    );
                    v['counter'] = i['counter'];
                    v['type'] = 'sysinclude';
                    ret.push(v);
                }
            });
            i['inputs'].forEach(v => {
                if (v['name'] === variable) {
                    v['iteration'] = self._exractIterationCoordinate(
                        i['iteration_coordinate']
                    );
                    v['counter'] = i['counter'];
                    v['type'] = 'input';
                    ret.push(v);
                }
            });
        });

        return ret;
    }

    /**
     * Get model viewer data
     *
     * @returns {Promise} resolves to model viewer data
     */
    async getModelViewerData() {
        return this._data.getModelViewerData();
    }

    /**
     * Get driver iteration data for a given variable if
     * new data is available based on the count given
     *
     * @param {string} variable variable name
     * @param {number} count highest current count
     * @returns {Promise}
     */
    async getDriverIterationsBasedOnCount(variable, count) {
        let ret = await this._data.isNewDriverIterationData(count);
        if (ret === true) {
            return this.getDriverIterationData(variable);
        }

        return [];
    }

    /**
     * Get the set of all variables recorded
     *
     * @returns {Promise}
     */
    async getAllDriverVars() {
        let dat = [];
        let ret = {
            desvars: [],
            objectives: [],
            constraints: [],
            sysincludes: [],
            inputs: []
        };

        try {
            dat = await this._data.getDriverIterationData();
        } catch (e) {
            return ret;
        }

        let cache = [];
        let self = this;
        dat.forEach(i => {
            i['desvars'].forEach(v => {
                let name = v['name'];
                if (!cache.includes(name)) {
                    ret.desvars.push(name);
                    cache.push(name);
                }
            });
            i['objectives'].forEach(v => {
                let name = v['name'];
                if (!cache.includes(name)) {
                    ret.objectives.push(name);
                    cache.push(name);
                }
            });
            i['constraints'].forEach(v => {
                let name = v['name'];
                if (!cache.includes(name)) {
                    ret.constraints.push(name);
                    cache.push(name);
                }
            });
            i['sysincludes'].forEach(v => {
                let name = v['name'];
                if (!cache.includes(name)) {
                    ret.sysincludes.push(name);
                    cache.push(name);
                }
            });
            i['inputs'].forEach(v => {
                let name = v['name'];
                if (!cache.includes(name)) {
                    ret.inputs.push(name);
                    cache.push(name);
                }
            });
        });

        return ret;
    }

    _exractIterationCoordinate(coord) {
        let delimiter = '|';
        let ret = [];

        let splitCoord = coord.split(delimiter);
        let i = 0;
        while (i < splitCoord.length) {
            let node = {
                name: splitCoord[i],
                iteration: splitCoord[i + 1]
            };
            ret.push(node);
            i += 2;
        }

        return ret;
    }
}

module.exports = Logic;
